Add support for _Float16 16-bit floating point type

jhendersonHDF opened the pull request:
This PR adds native support for a 16-bit floating point datatype through the `H5T_NATIVE_FLOAT16`, `H5T_IEEE_F16LE` and `H5T_IEEE_F16BE` macros. It is marked as a draft PR for early review while I'm still resolving issues with compilers that generate calls to functions like `__extendhfxf2` and then they end up not being properly linked in.

Things not added in this PR:

- No Fortran added yet; left for Scot though Fortran has no Float16 type. Should
  just make the identifiers available.
- No C++ added yet
- No `H5LTmake_dataset_float16` high-level method added yet since it would
  be a public API and whether or not _Float16 is available is not controllable
  by user. API would need to accept a const _Float16 *buffer parameter, where
  the type might not be available.
- Constants for types added to Java, but Java has no matching type yet
- No support for data transforms on _Float16 data added yet, because
  macro system is not setup to nicely support optional types
- No option to enable AVX512 FP16 instructions - user could add flag if desired

jhendersonHDF committed:
Add support for _Float16 16-bit floating point type

jhendersonHDF committed:
Fixed some _Float16 support issues

Fixed some conversion issues with Clang due to problematic undefined
behavior when casting a negative floating-point value to an integer

Fixed a bug in the library's software integer to floating-point
conversion function where a user's conversion exception function
returning H5T_CONV_UNHANDLED in the case of overflows would result in
incorrect data after conversion

Added configure checks for functions and macros related to _Float16
usage since some compilers expose the datatype but not the functions or
macros

Fixed a dt_arith test failure when H5_WANT_DCONV_EXCEPTION isn't defined

Fixed a few warnings from not explicitly casting some _Float16 variables
upwards

jhendersonHDF committed:
Add test program to see if _Float16 really works

jhendersonHDF committed:
Work around issues with MacOS 13

Add CMake status message about _Float16 test-compile program

jhendersonHDF committed:
Add RELEASE.txt note for _Float16 support

jhendersonHDF committed:
Merge branch 'develop' into feature/new_datatypes

jhendersonHDF commented on src/H5Tinit_float.c:
```diff
@@ -306,14 +306,17 @@ H5T__fix_order(int n, int last, int *perm, H5T_order_t *order)
-    else if (perm[last] > perm[last - 1] && perm[last - 1] > perm[last - 2]) {
+    else if (perm[last] > perm[last - 1] &&
+             /* Only check perm[last - 2] if we have more than 2 points to consider */
+             ((last < 2) || (perm[last - 1] > perm[last - 2]))) {
```
The changes here fix an issue where this function was assuming it could access `perm[last - 2]` even though that was invalid and would also cause the `_Float16` type to be detected as VAX endian order

jhendersonHDF commented on test/dt_arith.c:
```diff
@@ -601,19 +608,34 @@ static int
     int            endian; /*endianness            */
     unsigned int   fails_this_test = 0;
     size_t         j;
+#ifdef H5_WANT_DCONV_EXCEPTION
```
The changes in this function add `_Float16` testing and also skip running this part of the test when `H5_WANT_DCONV_EXCEPTION` is not defined, as there was otherwise a testing failure in that case since the library wouldn't raise the expected floating-point conversion exception.

jhendersonHDF commented on test/dt_arith.c:
```diff
@@ -2978,74 +3459,62 @@ test_conv_flt_1(const char *name, int run_test, hid_t src, hid_t dst)
 
         /* The hardware conversion */
-        /* Check for underflow when src is a "larger" float than dst.*/
-        if (FLT_FLOAT == src_type) {
```
At this point, it seemed reasonable to rewrite this part of dt_arith by creating separate functions for each type we're converting from. Otherwise, this started to become a visual mess and future support for additional types like complex numbers will only make it worse.

jhendersonHDF commented on src/H5Tnative.c:
```diff
@@ -715,7 +717,14 @@ H5T__get_native_float(size_t size, H5T_direction_t direction, size_t *struct_ali
     if (direction == H5T_DIR_DEFAULT || direction == H5T_DIR_ASCEND) {
-        if (size <= sizeof(float)) {
+#ifdef H5_HAVE__FLOAT16
+        if (size <= sizeof(H5__Float16)) {
```
It seems dangerous for the code here to match smaller datatypes with larger datatypes, since, before this change, the test code we currently have in ntypes.c was matching `_Float16` to `float` and causing an `H5Dread` call using the type returned from `H5Tget_native_type` to overrun the buffer. Seems to me that we should only match a type if the sizes match. That said, I'm just keeping in line with how the code behaved previously here.

fortnern replied:
I think that's a bug in the test code. `H5Tget_native_type` doesn't guarantee the size will match, it kind of fuzzily selects the closest native match to the datatype passed in. The app code should inspect the datatype returned to determine how to allocate the buffer and interpret the data. (If a real app knew the size a priori, it probably wouldn't need to call `H5Tget_native_type` anyways).

jhendersonHDF commented on src/H5Tconv.c:
```diff
@@ -7976,8 +8298,11 @@ H5T__conv_f_i(hid_t src_id, hid_t dst_id, H5T_cdata_t *cdata, size_t nelmts, siz
+            if (dst.prec / 8 > src_p->shared->size)
+                buf_size = (dst.prec + 7) / 8;
+            else
+                buf_size = src_p->shared->size;
```
This function was previously assuming ~32-bit floats. 16-bit and smaller floats would cause an invalid memory access further on when assuming it could copy more than 16 bits out of the temporary source buffer into the destination. In order to fix this bug, I had to rewrite parts of this function, which also led to a significant speedup when converting floats to ints in software.

The old way that this function used to work is that it would allocate a relatively large buffer to hold the temporary converted integer. The buffer would be `(2^exponent_num_bits / 8) + 1` bytes to try and have enough space so that the mantissa value can be shifted by any possible shift value from the exponent without losing bits from the buffer so that the MSB can be checked after the shift. However, for something like an IEEE 128-bit float with an exponent of 15 bits, this would allocate a 4097-byte buffer and then perform multiple different bit operations on all `4097 * 8 = 32776` bits of the buffer, which was very slow. Now, the function just allocates a buffer as big as the size of the destination integer and checks the MSB before shifting, then later on calculates the new MSB position after the shift to determine if overflow or underflow occurred. A small test program that converts 10,000,000 64-bit doubles to 32-bit ints went from 7.6 seconds to 1.05 seconds. Converting 10,000,000 IEEE 128-bit floats to 32-bit ints goes from 97.15 seconds to 1.17 seconds.

jhendersonHDF committed:
Update configure checks to account for problematic compilers

Clarify a few things in the RELEASE.txt entry

lrknox reviewed (approved):
The build errors we saw before on stria and perlmutter are gone now.

jhendersonHDF commented:
> The build errors we saw before on stria and perlmutter are gone now.

Great, thanks for testing!

qkoziol commented on tools/test/h5ls/expected/tfloat16.ls:
```diff
@@ -0,0 +1,8 @@
+Opened "tfloat16.h5" with sec2 driver.
+DS16BITS                 Dataset {8/8, 16/16}
+    Attribute: DS16BITS {128}
+        Type:      native _Float16
```
Will this be different on a big-endian machine?

jhendersonHDF replied:
Ah that's a good point. I should test this on a big-endian machine, but I suspect that yes it will probably trip over this test.

jhendersonHDF replied:
Added a BE test file for h5dump, h5dump xml and h5ls tests and verified testing on big-endian machine, but only in the case where `_Float16` isn't supported since we don't have a BE machine where it is. However, I setup the tests to only check the appropriate BE or LE test file, depending on the target architecture, for `native _Float16`, and verified the correct file is picked between BE and LE machines with both CMake and Autotools, so this should work on BE machines where `_Float16` is supported.

qkoziol commented on tools/test/h5dump/testh5dump.sh.in:
I think the tools/test/h5dump/testh5dumpxml.sh.in test script needs a test(s) for float16 also.

jhendersonHDF replied:
Seems reasonable, I'll add that.

jhendersonHDF replied:
Added an XML test. Also added one for the new BE file while I was at it.

qkoziol commented:
I don't know where the DDL definition file for h5dump output has gone, but it probably needs updating.

qkoziol commented:
Changes to the library and utilities (except some small things in h5dump) look good so far.

jhendersonHDF commented:
> I don't know where the DDL definition file for h5dump output has gone, but it probably needs updating.

Looks like the old DDL files are at https://github.com/HDFGroup/hdf5/blob/develop/doxygen/dox/DDLBNF110.dox and https://github.com/HDFGroup/hdf5/blob/develop/doxygen/dox/DDLBNF112.dox, but we don't seem to have one for the current release branch. I updated the part in src/H5Tmodule.h that references the datatypes portion of these though.

qkoziol commented:
> > I don't know where the DDL definition file for h5dump output has gone, but it probably needs updating.
> 
> Looks like the old DDL files are at https://github.com/HDFGroup/hdf5/blob/develop/doxygen/dox/DDLBNF110.dox and https://github.com/HDFGroup/hdf5/blob/develop/doxygen/dox/DDLBNF112.dox, but we don't seem to have one for the current release branch. I updated the part in src/H5Tmodule.h that references the datatypes portion of these though.

Good catch, that looks like it might hvae been missed in the doxygenation process.  :-)

byrnHDF commented:
Noone probably thought to update since 1.12 added the references?

I was likely the last to work on the DDL specs.

________________________________________
From: Quincey Koziol ***@***.***>
Sent: Sunday, March 10, 2024 10:04 AM
To: HDFGroup/hdf5
Cc: Allen Byrne; Review requested
Subject: Re: [HDFGroup/hdf5] Add support for _Float16 16-bit floating point type (PR #4065)

I don't know where the DDL definition file for h5dump output has gone, but it probably needs updating.

Looks like the old DDL files are at https://github.com/HDFGroup/hdf5/blob/develop/doxygen/dox/DDLBNF110.dox and https://github.com/HDFGroup/hdf5/blob/develop/doxygen/dox/DDLBNF112.dox, but we don't seem to have one for the current release branch. I updated the part in src/H5Tmodule.h that references the datatypes portion of these though.

Good catch, that looks like it might hvae been missed in the doxygenation process. :-)

—
Reply to this email directly, view it on GitHub<https://github.com/HDFGroup/hdf5/pull/4065#issuecomment-1987262452>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AL77KBS6AANFEK55XMKWACDYXSABDAVCNFSM6AAAAABEGGUTKKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOBXGI3DENBVGI>.
You are receiving this because your review was requested.Message ID: ***@***.***>

byrnHDF commented:
If there was no changes for DDL there would not be a need for a 1.14 version.

It's more likely that the 1.8 version is the one missing.

Allen

________________________________________
From: Allen Byrne ***@***.***>
Sent: Sunday, March 10, 2024 1:56 PM
To: HDFGroup/hdf5; HDFGroup/hdf5
Cc: Review requested
Subject: Re: [HDFGroup/hdf5] Add support for _Float16 16-bit floating point type (PR #4065)

Noone probably thought to update since 1.12 added the references?

I was likely the last to work on the DDL specs.

________________________________________
From: Quincey Koziol ***@***.***>
Sent: Sunday, March 10, 2024 10:04 AM
To: HDFGroup/hdf5
Cc: Allen Byrne; Review requested
Subject: Re: [HDFGroup/hdf5] Add support for _Float16 16-bit floating point type (PR #4065)

I don't know where the DDL definition file for h5dump output has gone, but it probably needs updating.

Looks like the old DDL files are at https://github.com/HDFGroup/hdf5/blob/develop/doxygen/dox/DDLBNF110.dox and https://github.com/HDFGroup/hdf5/blob/develop/doxygen/dox/DDLBNF112.dox, but we don't seem to have one for the current release branch. I updated the part in src/H5Tmodule.h that references the datatypes portion of these though.

Good catch, that looks like it might hvae been missed in the doxygenation process. :-)

—
Reply to this email directly, view it on GitHub<https://github.com/HDFGroup/hdf5/pull/4065#issuecomment-1987262452>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AL77KBS6AANFEK55XMKWACDYXSABDAVCNFSM6AAAAABEGGUTKKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOBXGI3DENBVGI>.
You are receiving this because your review was requested.Message ID: ***@***.***>

jhendersonHDF commented:
I can just go ahead and add a 1.14 DDL file since it will have these new types and eventually complex numbers.

jhendersonHDF committed:
Merge branch 'develop' into feature/new_datatypes

jhendersonHDF committed:
Address comments from review

Add 1.14 DDL file

jhendersonHDF committed:
Post-merge fixes

jhendersonHDF committed:
Add _Float16 h5dump XML test

Fix a few warnings in dt_arith.c test when _Float16 isn't available

jhendersonHDF committed:
Add big-endian _Float16 tools tests and files

jhendersonHDF commented:
> I don't know where the DDL definition file for h5dump output has gone, but it probably needs updating.

Added a DDLBNF114.dox file and updated references in the library to point to the new file.

derobins committed:
Merge branch 'develop' into feature/new_datatypes

derobins committed:
Merge branch 'develop' into feature/new_datatypes

jhendersonHDF committed:
Remove duplicate occurence of WORDS_BIGENDIAN in testh5ls.sh.in

fortnern reviewed (approved):
Library changes look good to me
